### PR TITLE
Adding support for multiple access token signing keys (CDI 2.9)

### PR DESCRIPTION
**Pull request by porcellus**
Adding support for multiple access token signing keys (CDI 2.9). This PR will smooth out the request spikes at signing key changes.

### CODE REVIEW

File: src/SuperTokens.AspNetCore/SuperTokensHandler.cs [hunk 1]
```diff
@@ -148,28 +148,42 @@ protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
...
+            AccessToken? parsedAccessToken = null;
+
+            // If we have no key old enough to verify this access token we should reject it without calling the core
+            var foundSigningKeyOlderThanToken = false;
             var isSignatureValid = false;
-            if (handshake.JwtSigningPublicKeyExpiration > now)
+            foreach (var keyInfo in handshake.AccessTokenSigningPublicKeyList)
             {
-                if (!JwtUtilities.TryParseAndValidate(accessToken, handshake.JwtSigningPublicKey, out var jwtPayload, out isSignatureValid) ||
+                if (!JwtUtilities.TryParseAndValidate(accessToken, keyInfo.PublicKey, out var jwtPayload, out isSignatureValid) ||
```
Comment (seniorquico) [inline src/SuperTokens.AspNetCore/SuperTokensHandler.cs]: There's already a `TryParse` method, and the `TryParseAndValidate` method was supposed to be an optimized routine for this specific use case. Parsing in a loop invalidates the optimization, and it's possible the `TryParseAndValidate` method doesn't make sense anymore. Maybe it'd be better to parse it just once and then, as an example, loop over each of the signing keys calling a `TryValidate` method?

File: src/SuperTokens.AspNetCore/HandshakeContainer.cs [hunk 2]
```diff
@@ -85,28 +97,38 @@ public async Task<Handshake> GetHandshakeAsync(string? apiKey, string? cdiVersio
...
+                        DateTimeOffset.FromUnixTimeMilliseconds(keyInfo.expirationTime),
+                        DateTimeOffset.FromUnixTimeMilliseconds(keyInfo.createdAt)
+                    ))
+                : new[] {
+                    new AccessTokenSigningKey(jwtSigningPublicKey,
+                        jwtSigningPublicKeyExpiration,
+                        DateTimeOffset.Now
+                )};
+
+            if (handshake.AccessTokenSigningPublicKeyList.Equals(updatedSigningPublicKeyList))
```
Comment (seniorquico) [inline src/SuperTokens.AspNetCore/HandshakeContainer.cs]: Isn't this checking for reference equality of the underlying array? (Same question applies to a few lines down from here.)

File: src/SuperTokens.AspNetCore/Handshake.cs [hunk 3]
```diff
@@ -1,31 +1,37 @@
...
+            this.AccessTokenSigningPublicKeyList = ImmutableArray.ToImmutableArray(jwtSigningPublicKeyList);
+
             this.AccessTokenBlacklistingEnabled = accessTokenBlacklistingEnabled;
             this.AccessTokenLifetime = accessTokenLifetime;
             this.RefreshTokenLifetime = refreshTokenLifetime;
         }
 
+        public ImmutableArray<AccessTokenSigningKey> AccessTokenSigningPublicKeyList
+        {
+            get => accessTokenSigningPublicKeyList.Where(keyInfo => keyInfo.Expiration > DateTimeOffset.Now).ToImmutableArray();
```
Comment (seniorquico) [inline src/SuperTokens.AspNetCore/Handshake.cs]: I know I don't have anything in here test-wise just yet... but from a testability perspective, the dependency on `DateTimeOffset.Now` makes this impossible to run over a static data set (say in a regression test suite). It also injects the requirement that certain tests be dependent on a real clock advancing as opposed to a simulated clock.

Two possible refactorings could eliminate the use of `DateTimeOffset.Now`:

1. Turn this into a method that accepts a "current time" argument used to filter the list. The caller would then be responsible for using an abstracted clock (e.g. `Microsoft.AspNetCore.Authentication.ISystemClock`) to get the current time and feed it into the method.
2. Return the array as-is and let the caller apply the filtering logic as needed. The caller would still be responsible for using an abstracted clock (e.g. `Microsoft.AspNetCore.Authentication.ISystemClock`) to get the current time for use in its own logic.

File: src/SuperTokens.AspNetCore/SuperTokensHandler.cs (hunk 1 above)
Comment (porcellus) [reply src/SuperTokens.AspNetCore/SuperTokensHandler.cs]: I split it up into multiple steps and moved all parsing outside the loop.

**Review by seniorquico — COMMENTED** [review]
Awesome!

**Comment by porcellus** [comment]
These charts show requests per second to different endpoints during the same tests run on different core and C# backend versions. The title of each graph indicates different things:
- Starting with `browser` are the calls made from the browser to the backend SDK
- Starting with `node` are the calls made from the backend SDK to the core
- `csharp_main` means it used the main branch, while `csharp_multi` means the branch of this PR
- after `__` comes the branch of the core used during the test: 
  - `multi` once again means the core with the modifications necessary for this PR
  - `3_6` means one without.

image
image
image
image
image
image
image
image

**Comment by seniorquico** [comment]
That shows a 15x drop in peak calls to the Core, right? This should dramatically decrease the Core's load! Awesome improvement. If I understand it correctly, the security implications/trade-offs of this change are that we will keep signing keys longer than before (so that there's overlap from one to the next). Are there any other implications/trade-offs? Are there any plans to run an analysis to try and identify a "sweet spot" for the validity period?

**Comment by porcellus** [comment]
Well, the 15x drop is unlikely to happen in a real-world scenario, the API usage in the test is much more consistent than any real user. What I'd expect is that the call spikes will drop back to the normal level. This behavior was caused by the signing key change requiring refresh from all active sessions at once, basically bunching up the traffic. You can see this in the last chart: there is some normal traffic that spikes then drops to zero and spikes again.
We keep the signing keys for longer (signing key update interval + 2* access token validity). However, we are still checking the token expiration even if we accept the signature with the "old" keys. It means that the access tokens will be valid for the configured amount of time instead of ending early with the signing key update.
Other implications:
- Stolen DB: a stolen signing key is useful for 2 more access tokens (per user impersonated)
- Stolen access token: (stolen) tokens are valid for the same amount of time as before the change, except in edge cases when the token validity was cut short by the signing key update.

I'm not sure about plans to do a more in-depth analysis, but I think it would depend on your usage patterns. 
- Access token validity: I'd suggest using an access token validity that matches with typical time users will use your app "in one sitting" plus some margin. 
- Signing key update period: anything lower than twice the access token validity isn't really helpful, but it shouldn't cause much overhead if it isn't extremely short (e.g., 10 seconds). So 1 day seems like a good choice.

**Comment by seniorquico** [comment]
Yes, we also see the spikes due to the "bunching" and it takes several hours for it to return to normal:

image

**Review by seniorquico — APPROVED** [review]

**Comment by seniorquico** [comment]
@porcellus Sorry, this conflicted with the other PR. Ready for it to get in as soon as it's resolved!

**Review by seniorquico — APPROVED** [review]

**Pull request merged**